pkg/client/workspace.go

```diff
@@ -0,0 +1,283 @@
+		}
+		w.LastOperation.MarkEnd()
+		w.logger.Debug("destroy async completed", "stdout", stdout.String())
+		w.logger.Debug("destroy async completed", "stderr", stderr.String())
+
+		// After the operation is completed, we need to get the results saved on
+		// the custom resource as soon as possible. We can wait for the next
+		// reconcilitaion, enqueue manually or update the CR independent of the
+		// reconciliation.
+		w.Enqueue()
+		cancel()
+	}()
+	return nil
+}
+
+func (w *Workspace) Destroy(ctx context.Context) error {
+	if w.LastOperation.EndTime == nil {
+		return errors.Errorf("%s operation that started at %s is still running", w.LastOperation.Type, w.LastOperation.StartTime.String())
+	}
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+	cmd := exec.CommandContext(ctx, "terraform", "destroy", "-auto-approve", "-input=false", "-json")
+	cmd.Dir = w.dir
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+	if err := cmd.Run(); err != nil {
+		w.LastOperation.Err = errors.Wrapf(err, "cannot destroy: %s", stderr.String())
+	}
+	w.logger.Debug("destroy completed", "stdout", stdout.String())
+	w.logger.Debug("destroy completed", "stderr", stderr.String())
+	return nil
+}
+
+type RefreshResult struct {
+	IsApplying         bool
+	IsDestroying       bool
+	State              *json.StateV4
+	LastOperationError error
+}
+
+func (w *Workspace) Refresh(ctx context.Context) (RefreshResult, error) {
+	if w.LastOperation.StartTime != nil {
+		// The last operation is still ongoing.
+		if w.LastOperation.EndTime == nil {
+			return RefreshResult{
+				IsApplying:   w.LastOperation.Type == "apply",
+				IsDestroying: w.LastOperation.Type == "destroy",
+			}, nil
+		}
+		// We know that the operation finished, so we need to flush so that new
+		// operation can be started.
+		defer w.LastOperation.Flush()
+
+		// The last operation finished with error.
+		if w.LastOperation.Err != nil {
+			return RefreshResult{
+				IsApplying:         w.LastOperation.Type == "apply",
+				IsDestroying:       w.LastOperation.Type == "destroy",
+				LastOperationError: errors.Wrapf(w.LastOperation.Err, "%s operation failed", w.LastOperation.Type),
+			}, nil
+		}
+		// The deletion is completed so there is no resource to refresh.
+		if w.LastOperation.Type == "destroy" {
+			return RefreshResult{}, kerrors.NewNotFound(schema.GroupResource{}, "")
+		}
+	}
+	stdout := &bytes.Buffer{}
+	stderr := &bytes.Buffer{}
+	cmd := exec.CommandContext(ctx, "terraform", "apply", "-refresh-only", "-auto-approve", "-input=false", "-json")
+	cmd.Dir = w.dir
+	cmd.Stdout = stdout
+	cmd.Stderr = stderr
+	// In case the resource does not exist, this command doesn't return an error.
+	// It only removes the resource from tfstate file. We need to call Plan to
+	// get an idea about whether the resource exists from Terraform's perspective.
+	if err := cmd.Run(); err != nil {
+		return RefreshResult{}, errors.Wrapf(err, "cannot refresh: %s", stderr.String())
```

Comment:
I think this shortcut somewhat deviates from other Crossplane providers' usual pattern of request a deletion, and then observe the external resource (the world state) until it's reported to be missing by the Cloud API.

Reply:
Do you mean the case where we report that resource does not exist if `terraform` completed the destroy operation successfully?

Reply:
If so, this is still the case in controller level (pushed a change to integrate with controller) so it's still the same logic from managed reconciler perspective. But we do take a shortcut here at the client level saying resource does not exist if terraform destroy is successful. Theoretically, it's true that resource may still exist (like in PostgreSQLConfiguration case where the deletion is just setting back to default, or KMS keys where deletion is completed after 14 days) even though the destroy is completed. However, I think it's worth relying on terraform here to reuse their logic around that. Otherwise, we'd have to add custom logic for such resources to approve the deletion.